fs/localfs/copy.go

```diff
@@ -0,0 +1,164 @@
+// to the root directory.
+// Copy does not overwrite files or directories and returns an error in that
+// case. It also returns an error when the the contents cannot be restored,
+// for example due to an I/O error.
+func Copy(ctx context.Context, targetPath string, e fs.Entry) error {
+	targetPath, err := filepath.Abs(filepath.FromSlash(targetPath))
+	if err != nil {
+		return err
+	}
+
+	root, err := filepath.Abs(filepath.FromSlash("/"))
+	if err != nil {
+		return err
+	}
+
+	// special case when the target is the root directory
+	if targetPath == root {
+		if d, ok := e.(fs.Directory); ok {
+			return copyDirectoryContent(ctx, d, targetPath)
+		}
+
+		return errors.Errorf("cannot restore non-directory to root target path %q: %#v", targetPath, e)
+	}
+
+	return copyEntry(ctx, e, targetPath)
+}
+
+func copyEntry(ctx context.Context, e fs.Entry, targetPath string) error {
+	var err error
+
+	switch e := e.(type) {
+	case fs.Directory:
+		err = copyDirectory(ctx, e, targetPath)
+	case fs.File:
+		err = copyFileContent(ctx, targetPath, e)
+	case fs.Symlink:
+		// Not yet implemented
+		log.Warningf("Not creating symlink %q from %v", targetPath, e)
+		return nil
+	default:
+		return errors.Errorf("invalid FS entry type for %q: %#v", targetPath, e)
+	}
+
+	if err != nil {
+		return err
+	}
+
+	return setAttributes(targetPath, e)
+}
+
+// set permission, modification time and user/group ids on targetPath
+func setAttributes(targetPath string, e fs.Entry) error {
+	const modBits = os.ModePerm | os.ModeSetgid | os.ModeSetuid | os.ModeSticky
+
+	le, err := NewEntry(targetPath)
+	if err != nil {
+		return errors.Wrap(err, "could not create local FS entry for "+targetPath)
+	}
+
+	// Set owner user and group from e
+	if le.Owner() != e.Owner() {
+		if err = os.Chown(targetPath, int(e.Owner().UserID), int(e.Owner().GroupID)); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change owner/group for "+targetPath)
+		}
+	}
+
+	// Set file permissions from e
+	if (le.Mode() & modBits) != (e.Mode() & modBits) {
+		if err = os.Chmod(targetPath, e.Mode()&modBits); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change permissions on "+targetPath)
+		}
+	}
+
+	// Set mod time from e
+	if !le.ModTime().Equal(e.ModTime()) {
+		// Note: Set atime to ModTime as well
+		if err = os.Chtimes(targetPath, e.ModTime(), e.ModTime()); err != nil && !os.IsPermission(err) {
+			return errors.Wrap(err, "could not change mod time on "+targetPath)
+		}
+	}
+
+	return nil
+}
+
+func copyDirectory(ctx context.Context, d fs.Directory, targetPath string) error {
+	if err := createDirectory(targetPath); err != nil {
+		return err
+	}
+
+	return copyDirectoryContent(ctx, d, targetPath)
+}
+
+func copyDirectoryContent(ctx context.Context, d fs.Directory, targetPath string) error {
+	entries, err := d.Readdir(ctx)
+	if err != nil {
+		return err
+	}
+
+	for _, e := range entries {
```

Comment:
for high-latency repositories this will be pretty slow unless we parallelize. Fortunately there's `parallelwork.Queue` which can be used to pretty efficiently schedule work whose size is unknown when you begin. Basically - instead of just doing work in the for loop just schedule each entry in the queue and let the queue do the magic.